Bug:  Failed to keep quota

### What happened?

I'm writing data to a bucket with the FIFO quota continuously. Sometimes, the write operation fails with the following error:

```
2024-11-21 15:08:58.375 (35072) [ERROR] -- api/middleware.rs:56 POST /api/v1/b/stress_test/entry-1?ts=1732201738263914 [500 Internal Server Error] Failed to keep quota of 'stress_test'
```

### Version

v1.13.0-dev

### OS

Linux

### Snippet to repeat the bug

_No response_

### Logs

```Shell
2024-11-20 19:26:17.401 (35072) [INFO] -- main.rs:23 ReductStore 1.12.4 [6966266 at 2024-11-20T19:17:52Z]
2024-11-20 19:26:17.420 (35072) [INFO] -- main.rs:38 Configuration: 
 	RS_LOG_LEVEL = INFO 
	RS_HOST = 0.0.0.0 (default)
	RS_PORT = 8383 (default)
	RS_API_BASE_PATH = / (default)
	RS_DATA_PATH = /data (default)
	RS_API_TOKEN = *********** 
	RS_BUCKET_2_NAME = repl 
	RS_BUCKET_3_NAME = stress_test 
	RS_BUCKET_2_QUOTA_TYPE = FIFO 
	RS_BUCKET_2_QUOTA_SIZE = 5.0 GB 
	RS_BUCKET_3_QUOTA_TYPE = FIFO 
	RS_BUCKET_3_QUOTA_SIZE = 10.0 GB 
	RS_TOKEN_2_NAME = repl-rw 
	RS_TOKEN_3_NAME = stress_test-rw 
	RS_TOKEN_2_VALUE = ********************************* 
	RS_TOKEN_3_VALUE = *********************************** 
	RS_TOKEN_2_READ = repl 
	RS_TOKEN_2_WRITE = repl 
	RS_TOKEN_3_READ = stress_test 
	RS_TOKEN_3_WRITE = stress_test 
	RS_TOKEN_3_FULL_ACCESS = false 
	RS_REPLICATION_1_NAME = stress_test 
	RS_REPLICATION_1_SRC_BUCKET = stress_test 
	RS_REPLICATION_1_DST_BUCKET = repl 
	RS_REPLICATION_1_DST_HOST = http://reductstore:8383 
	RS_REPLICATION_1_DST_TOKEN = ********************************* 
	RS_REPLICATION_1_INCLUDE_size = medium 

2024-11-20 19:26:17.431 (85344) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/stress_test/entry-2"
2024-11-20 19:26:17.431 (91040) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/stress_test/entry-1"
2024-11-20 19:26:17.561 (85344) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/repl/entry-1"
2024-11-20 19:26:17.561 (91040) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/repl/entry-2"
2024-11-20 19:26:17.701 (35072) [INFO] -- storage/storage.rs:67 Load 2 buckets
2024-11-20 19:26:17.702 (35072) [INFO] -- reductstore/src/cfg.rs:146 Provisioned bucket 'repl' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(5000000000), max_block_size: Some(64000000), max_block_records: Some(1024) }
2024-11-20 19:26:17.702 (35072) [INFO] -- reductstore/src/cfg.rs:146 Provisioned bucket 'stress_test' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(10000000000), max_block_size: Some(64000000), max_block_records: Some(1024) }
2024-11-20 19:26:17.703 (35072) [INFO] -- reductstore/src/cfg.rs:112 Provisioned token 'stress_test-rw' with Some(Permissions { full_access: false, read: ["stress_test"], write: ["stress_test"] })
2024-11-20 19:26:17.703 (35072) [INFO] -- reductstore/src/cfg.rs:112 Provisioned token 'repl-rw' with Some(Permissions { full_access: false, read: ["repl"], write: ["repl"] })
2024-11-20 19:26:17.703 (35072) [INFO] -- cfg.rs:372 Load Web Console
2024-11-20 19:26:18.035 (35072) [INFO] -- reductstore/src/cfg.rs:176 Provisioned replication 'stress_test' with ReplicationSettings { src_bucket: "stress_test", dst_bucket: "repl", dst_host: "http://reductstore:8383/", dst_token: "***", entries: [], include: {"size": "medium"}, exclude: {}, each_n: None, each_s: None }
2024-11-20 19:26:18.035 (35072) [INFO] -- main.rs:48 License: BUSL-1.1 [https://github.com/reductstore/reductstore/blob/v1.12.4/LICENSE]
2024-11-20 19:26:18.035 (35072) [INFO] -- main.rs:60 Run HTTP reductstore on http://0.0.0.0:8383/
2024-11-20 19:26:18.044 (19520) [ERROR] -- replication/remote_bucket/states/initial_state.rs:50 Failed to get remote bucket http://reductstore:8383/repl: [ConnectionError] error sending request for url (http://reductstore:8383/api/v1/b/repl)
2024-11-20 19:26:18.044 (36608) [ERROR] -- replication/remote_bucket/states/initial_state.rs:50 Failed to get remote bucket http://reductstore:8383/repl: [ConnectionError] error sending request for url (http://reductstore:8383/api/v1/b/repl)
2024-11-21 15:08:58.375 (35072) [ERROR] -- api/middleware.rs:56 POST /api/v1/b/stress_test/entry-1?ts=1732201738263914 [500 Internal Server Error] Failed to keep quota of 'stress_test'
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Update Web Console up to v1.8.0, [PR-655](https://github.com/reductstore/reductstore/pull/655)
 
+
+### Fixed
+
+- RS-544: Fix keeping quota error, [PR-654](https://github.com/reductstore/reductstore/pull/654)
+
 ### Internal
 
 - Fix CI actions after Ubuntu update, [PR-604](https://github.com/reductstore/reductstore/pull/604)

--- a/reductstore/src/core/thread_pool.rs
+++ b/reductstore/src/core/thread_pool.rs
@@ -12,7 +12,7 @@ use std::num::NonZero;
 use std::ops::Deref;
 use std::path::PathBuf;
 use std::sync::{Arc, LazyLock, Mutex};
-use std::thread::{available_parallelism, JoinHandle};
+use std::thread::{available_parallelism, sleep, JoinHandle};
 use std::time::Duration;
 pub(crate) use task_group::TaskGroup;
 pub(crate) use task_handle::TaskHandle;
@@ -84,17 +84,22 @@ where
 pub(crate) fn try_unique<T>(
     group_path: &str,
     description: &str,
+    timeout: Duration,
     task: impl FnOnce() -> T + Send + 'static,
 ) -> Option<TaskHandle<T>>
 where
     T: Send + 'static,
 {
     let path = group_path.split('/').collect::<Vec<&str>>();
-    if THREAD_POOL.task_group.lock().unwrap().is_ready(&path, true) {
-        Some(unique(group_path, description, task))
-    } else {
-        None
+    let now = std::time::Instant::now();
+    while now.elapsed() < timeout {
+        if THREAD_POOL.task_group.lock().unwrap().is_ready(&path, true) {
+            return Some(unique(group_path, description, task));
+        }
+        sleep(Duration::from_micros(10));
     }
+
+    None
 }
 
 pub(crate) enum GroupDepth {

--- a/reductstore/src/storage/block_manager.rs
+++ b/reductstore/src/storage/block_manager.rs
@@ -6,7 +6,7 @@ pub(in crate::storage) mod block_index;
 pub(in crate::storage) mod wal;
 
 mod block_cache;
-use log::{debug, error, trace};
+use log::{debug, error, trace, warn};
 use prost::bytes::{Bytes, BytesMut};
 use prost::Message;
 
@@ -15,7 +15,7 @@ use crate::storage::block_manager::block::Block;
 use crate::storage::block_manager::block_cache::BlockCache;
 use crate::storage::block_manager::wal::{Wal, WalEntry};
 use crate::storage::entry::io::record_reader::read_in_chunks;
-use crate::storage::proto::{record, ts_to_us, Block as BlockProto, Record};
+use crate::storage::proto::{record, ts_to_us, us_to_ts, Block as BlockProto, Record};
 use crate::storage::storage::IO_OPERATION_TIMEOUT;
 use block_index::BlockIndex;
 use crc64fast::Digest;
@@ -130,9 +130,18 @@ impl BlockManager {
             }
 
             // parse the block descriptor
-            let block_from_disk = BlockProto::decode(Bytes::from(buf)).map_err(|e| {
+            let mut block_from_disk = BlockProto::decode(Bytes::from(buf)).map_err(|e| {
                 internal_server_error!("Failed to decode block descriptor {:?}: {}", path, e)
             })?;
+
+            if block_from_disk.begin_time.is_none() {
+                warn!(
+                    "Block descriptor {:?} has no begin time. It might be recovered from the WAL",
+                    path
+                );
+                block_from_disk.begin_time = Some(us_to_ts(&block_id));
+            }
+
             cached_block = Some(Arc::new(RwLock::new(block_from_disk.into())));
         }
 

--- a/reductstore/src/storage/block_manager.rs
+++ b/reductstore/src/storage/block_manager.rs
@@ -650,7 +650,7 @@ mod tests {
 
             let result = block_manager.load_block(block_id);
             assert!(
-                result.ok(),
+                result.is_ok(),
                 "It's ok to recover begin time from block id for blocks which aren't synced yet"
             );
         }

--- a/reductstore/src/storage/block_manager.rs
+++ b/reductstore/src/storage/block_manager.rs
@@ -641,6 +641,21 @@ mod tests {
         }
 
         #[rstest]
+        fn test_recover_being_time_from_id(mut block_manager: BlockManager, block_id: u64) {
+            block_manager.start_new_block(block_id, 1024).unwrap();
+            block_manager.block_cache.remove(&block_id);
+
+            let path = block_manager.path_to_desc(block_id);
+            std::fs::write(&path, b"").unwrap();
+
+            let result = block_manager.load_block(block_id);
+            assert!(
+                result.ok(),
+                "It's ok to recover begin time from block id for blocks which aren't synced yet"
+            );
+        }
+
+        #[rstest]
         fn test_start_reading(mut block_manager: BlockManager, block_id: u64) {
             let block = block_manager.start_new_block(block_id, 1024).unwrap();
             let block_id = block.read().unwrap().block_id();

--- a/reductstore/src/storage/bucket/quotas.rs
+++ b/reductstore/src/storage/bucket/quotas.rs
@@ -3,7 +3,7 @@
 
 use crate::storage::bucket::Bucket;
 use crate::storage::entry::Entry;
-use log::debug;
+use log::{debug, info, warn};
 use reduct_base::error::ReductError;
 use reduct_base::msg::bucket_api::QuotaType;
 use reduct_base::{bad_request, internal_server_error};
@@ -55,7 +55,7 @@ impl Bucket {
                         break;
                     }
                     Err(e) => {
-                        debug!("Failed to remove oldest block from entry '{}': {}", name, e);
+                        warn!("Failed to remove oldest block from entry '{}': {}", name, e);
                     }
                 }
             }

--- a/reductstore/src/storage/bucket/quotas.rs
+++ b/reductstore/src/storage/bucket/quotas.rs
@@ -3,7 +3,7 @@
 
 use crate::storage::bucket::Bucket;
 use crate::storage::entry::Entry;
-use log::{debug, info, warn};
+use log::{debug, warn};
 use reduct_base::error::ReductError;
 use reduct_base::msg::bucket_api::QuotaType;
 use reduct_base::{bad_request, internal_server_error};

--- a/reductstore/src/storage/entry.rs
+++ b/reductstore/src/storage/entry.rs
@@ -22,7 +22,7 @@ use std::fs;
 use std::path::PathBuf;
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::{Arc, RwLock};
-use std::time::Instant;
+use std::time::{Duration, Instant};
 use tokio::sync::RwLock as AsyncRwLock;
 
 pub(crate) use io::record_writer::{RecordDrainer, RecordWriter, WriteRecordContent};
@@ -226,6 +226,7 @@ impl Entry {
         match try_unique(
             &format!("{}/{}", self.task_group(), oldest_block_id),
             "try remove oldest block",
+            Duration::from_millis(5),
             move || {
                 let mut bm = block_manager.write().unwrap();
                 bm.remove_block(oldest_block_id)?;

--- a/reductstore/src/storage/entry.rs
+++ b/reductstore/src/storage/entry.rs
@@ -225,7 +225,7 @@ impl Entry {
 
         match try_unique(
             &format!("{}/{}", self.task_group(), oldest_block_id),
-            "try remove oldest block",
+            "try to remove oldest block",
             Duration::from_millis(5),
             move || {
                 let mut bm = block_manager.write().unwrap();

